File: pet-store-service/src/main/java/com/chewy/osn/service/PetService.java
package com.chewy.osn.service;

import com.chewy.osn.domain.Pet;
import com.chewy.osn.domain.Species;
import com.chewy.osn.exception.PetException;
import com.chewy.osn.repository.PetRepository;
import com.chewy.osn.response.CutenessResponse;
import com.chewy.osn.response.GetPetResponse;
import jakarta.inject.Inject;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class PetService {

    private PetRepository petRepository;

    @Inject
    public PetService(PetRepository petRepository) {
        this.petRepository = petRepository;
    }

    public GetPetResponse getPet(Species species, String name) {
        Pet pet = petRepository.getPet(species, name);

        if (pet == null) {
            throw new PetException(404, name + " was already adopted! :D");
        }
        return new GetPetResponse(pet.getCuteness());
    }

    public CutenessResponse getTotalCuteness(Species species, List<String> names) {
        List<Pet> pets = getPets(species, names);
        Integer totalCuteness = calculateCuteness(pets);

        return new CutenessResponse(species, names, totalCuteness);
    }

    public Integer calculateCuteness(List<Pet> cutePets) {
        return cutePets.stream()
            .map(Pet::getCuteness)
            .filter(Objects::nonNull)
            .mapToInt(Integer::intValue).sum();
    }

    public List<Pet> getPets(Species species, List<String> names) {
        List<Pet> pets = new ArrayList<>();

        for (String name: names) {
            Pet retrieved = petRepository.getPet(species, name);

            if (retrieved != null) {
                pets.add(retrieved);
            }
        }
        return pets;
    }

    public void createPet(Species species, String name, Integer cuteness) {
        Pet existingPet = petRepository.getPet(species, name);

        if (existingPet != null) {
            throw new PetException(409, "We already have a " + species + " that's named " + name);
        }

        petRepository.writePet(new Pet(name, species, cuteness));
    }
}

File: pet-store-service/src/main/kotlin/com/johnkruss/osn/service/PetService.kt
package com.johnkruss.osn.service

import com.johnkruss.osn.domain.Pet
import com.johnkruss.osn.domain.Species
import com.johnkruss.osn.dynamo.PetRepository
import com.johnkruss.osn.exception.PetException
import com.johnkruss.osn.response.CutenessResponse
import com.johnkruss.osn.response.GetPetResponse
import jakarta.inject.Inject

class PetService
    @Inject
    constructor(private val petRepository: PetRepository) {
        fun getPet(
            species: Species,
            name: String,
        ): GetPetResponse {
            val pet = petRepository.getPet(species, name) ?: throw PetException(404, "$name was already adopted! :D")

            return GetPetResponse(pet.cuteness)
        }

        fun getTotalCuteness(
            species: Species,
            names: List<String>,
        ): CutenessResponse {
            val pets = getPets(species, names)
            val totalCuteness = calculateCuteness(pets)

            return CutenessResponse(species, names, totalCuteness)
        }

        fun calculateCuteness(cutePets: List<Pet>): Int {
            return cutePets.mapNotNull { it.cuteness }.sum()
        }

        fun getPets(
            species: Species,
            names: List<String>,
        ): List<Pet> {
            val pets: MutableList<Pet> = ArrayList()

            for (name in names) {
                val retrieved = petRepository.getPet(species, name)

                if (retrieved != null) {
                    pets.add(retrieved)
                }
            }
            return pets
        }

        fun createPet(
            species: Species,
            name: String,
            cuteness: Int,
        ) {
            val existingPet = petRepository.getPet(species, name)

            if (existingPet != null) {
                throw PetException(409, "We already have a $species that's named $name")
            }

            petRepository.writePet(Pet(name, species, cuteness))
        }
    }
